### PR TITLE
Fix bulk archive bug

### PR DESCRIPTION
**Pull request by dannyroberts**
##### SUMMARY
Fixes https://dimagi-dev.atlassian.net/browse/SAASP-10215

This may not seem like much, but it actually took me quite a while to figure out how to address the original issue, and then almost as long to figure out what things I needed to set up to make the test pass after the fix—and fail before the fix.

The overall issue was that an internal call that previously returned a list of xforms (the ones to archive) is now returning an HttpResponse object for the “Sorry, you do not have access to Project Features” page. At first, my instinct was to rewrite this hacky call, which essentially mocks a request to the helper object that manages the page the user was just on, using a more sane function call, but that turned out to be surprisingly difficult, and I tried a few different ways. In order to make sure the behavior is really the same as what the user just saw, mocking a request to the same helper object seems to be the only viable option, barring a bigger rewrite.

The test I put in place should fail or error if another future permissions change triggers something like this in the future.

##### PRODUCT DESCRIPTION
Bulk Archive was broken if you selected the checkbox to archive (or unarchive) all forms in a search.

### CODE REVIEW

File: corehq/apps/data_interfaces/views.py [hunk 1]
```diff
@@ -518,6 +518,7 @@ def get_xform_ids(self, request):
             _request.couch_user.current_domain = self.domain
             _request.can_access_all_locations = request.couch_user.has_permission(self.domain,
                                                                                   'access_all_locations')
+            _request.always_allow_project_access = True
```
Comment (millerdev) [inline corehq/apps/data_interfaces/views.py]: This seems kinda magic, and possibly fragile. Does it override the permission check immediately above?

**Review by millerdev — APPROVED** [review]
The way `get_xform_ids` is constructing and dispatching a request seems like it could be vulnerable to all sorts of bad outcomes related to unintended HTTP responses similar to the one that caused the issue. Does that sound right?

File: corehq/apps/data_interfaces/views.py (hunk 1 above)
Comment (dannyroberts) [reply corehq/apps/data_interfaces/views.py]: Yeah it's totally magic and totally fragile. Specifically about the permissions check, "project access" just means allowed to access the project at all; it doesn't give blanket access to everything in the project.

**Comment by dannyroberts** [comment]
Yeah that's right, so any future change to permissions checks that affect the target interface could result in similar bugs to the one I'm trying to fix. My hope is that the test I added would catch that; for example that test (had it already been in the codebase) would have failed on the PR that most recently introduced this issue.

I did think of multiple different ways I could rewrite this more cleanly, but none of the reasonable effort ones were clearly better. The thing that makes it hard is all of the magic that makes the code surrounding parsing variables out of URLs and related behavior encapsulated and also a tangled mess so as to make it completely inaccessible.

**Comment by dannyroberts** [comment]
Yeah that's right, so any future change to permissions checks that affect the target interface could result in similar bugs to the one I'm trying to fix. My hope is that the test I added would catch that; for example that test (had it already been in the codebase) would have failed on the PR that most recently introduced this issue.

I did think of multiple different ways I could rewrite this more cleanly, but none of the reasonable effort ones were clearly better. The thing that makes it hard is all of the magic that makes the code surrounding parsing variables out of URLs and related behavior encapsulated and also a tangled mess so as to make it completely inaccessible.

**Pull request merged**